File: src/main/java/net/maple/handlers/user/UserTransferChannelRequestHandler.java
package net.maple.handlers.user;

import client.Character;
import client.Client;
import net.maple.SendOpcode;
import net.maple.handlers.PacketHandler;
import net.server.ChannelServer;
import net.server.Server;
import util.packet.Packet;
import util.packet.PacketReader;
import util.packet.PacketWriter;

public class UserTransferChannelRequestHandler extends PacketHandler {

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        Character chr = c.getCharacter();

        byte channelId = reader.readByte();
        // todo field limit check

        try {
            ChannelServer channel = Server.getInstance().getChannels().get(channelId);
            if (channel != null) {
                if (!c.getWorldChannel().equals(channel)) {
                    c.acquireMigrateState();
                    try {
                        //c.getCharacter().save();
                        c.changeChannel(channel);
                    } finally {
                        c.releaseMigrateState();
                    }
                    return;
                }
                c.write(fail());
                //c.close(this, "CC to same channel");
            } else {
                System.err.println("[UserTransferChannelRequestHandler] Channel is null (" + chr + ")");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        c.write(fail());
    }

    private static Packet fail() {
        PacketWriter pw = new PacketWriter(3);

        pw.writeHeader(SendOpcode.TRANSFER_CHANNEL_REQ_IGNORED);
        pw.write(0x01);

        return pw.createPacket();
    }
}

File: src/main/java/net/maple/handlers/user/UserTransferChannelRequestHandler.kt
package net.maple.handlers.user

import client.Client
import net.maple.SendOpcode
import net.maple.handlers.PacketHandler
import net.server.Server.channels
import util.logging.LogType
import util.logging.Logger.log
import util.packet.Packet
import util.packet.PacketReader
import util.packet.PacketWriter

class UserTransferChannelRequestHandler : PacketHandler {

    override fun handlePacket(reader: PacketReader, c: Client) {
        val chr = c.character
        val channelId = reader.readByte()
        // todo field limit check
        val id = channelId.toInt()
        try {
            val channel = channels[id]

            if (c.worldChannel != channel) {
                c.acquireMigrateState()
                try {
                    //c.getCharacter().save();
                    c.changeChannel(channel)
                } finally {
                    c.releaseMigrateState()
                }
                return
            }
            //c.write(fail())
            //c.close(this, "CC to same channel");
        } catch (e: Exception) {
            log(LogType.NULL, "Channel is null ($id) (${channels.size})", this, c)
            e.printStackTrace()
        }
        c.write(fail())
    }

    companion object {
        private fun fail(): Packet {
            val pw = PacketWriter(3)

            pw.writeHeader(SendOpcode.TRANSFER_CHANNEL_REQ_IGNORED)
            pw.write(0x01)

            return pw.createPacket()
        }
    }
}